internal/notification/slack/slack.go

```diff
@@ -0,0 +1,184 @@
+		tables = append(tables, []string{statusText(ts.Passed), regenerateName(ts.Name, r.getJobURL(ts.Name, ts.URL), longestName),
+			ts.Platform, ts.DeviceName, ts.Browser, ts.Duration.Truncate(1 * time.Second).String()})
+		if !ts.Passed {
+			passed = false
+		}
+	}
+	var res string
+	for _, t := range tables {
+		res = fmt.Sprintf("%s\n%s", res, strings.Join(t, "\t"))
+	}
+
+	r.RenderedResult = res
+
+	r.sendMessage(passed)
+}
+
+// GetRenderedResult returns rendered result.
+func (r *Reporter) GetRenderedResult() string {
+	return r.RenderedResult
+}
+
+// Reset no need to implement
+func (r *Reporter) Reset() {}
+
+// ArtifactRequirements no need to implement
+func (r *Reporter) ArtifactRequirements() []report.ArtifactType {
+	return nil
+}
+
+// sendMessage sends notification message.
+func (r *Reporter) sendMessage(passed bool) {
+	if !r.shouldSendNotification(passed) {
+		return
+	}
+
+	token, err := r.Service.GetSlackToken(context.Background())
+	if err != nil {
+		log.Err(err).Msg("Failed to get slack token")
+		return
+	}
+
+	api := slack.New(token)
+
+	for _, c := range r.Channels {
+		_, _, err := api.PostMessage(
+			c,
+			slack.MsgOptionText("saucectl test result", false),
+			slack.MsgOptionAttachments(r.creatAttachment(passed)),
+			slack.MsgOptionAsUser(true),
+		)
+
+		if err != nil {
+			log.Error().Msgf("Failed to send message to slack, err: %s", err.Error())
+			continue
+		}
+
+		log.Info().Msgf("Message successfully sent to slack channel %s", c)
+	}
+}
+
+// shouldSendNotification returns true if it should send notification, otherwise false
+func (r *Reporter) shouldSendNotification(passed bool) bool {
+	if len(r.Config.Slack.Channels) == 0 ||
+		r.Config.Slack.Send == config.WhenNever {
+		return false
+	}
+
+	if r.Config.Slack.Send == config.WhenAlways ||
+		(r.Config.Slack.Send == config.WhenFail && !passed) ||
+		(passed && r.Config.Slack.Send == config.WhenPass) {
+		return true
+	}
+
+	return false
+}
+
+func (r *Reporter) createBlocks() []slack.Block {
+	headerText := slack.NewTextBlockObject("mrkdwn", fmt.Sprintf("*%s*", r.Metadata.Build), false, false)
+	headerSection := slack.NewSectionBlock(headerText, nil, nil)
+
+	contextText := slack.NewTextBlockObject("mrkdwn", fmt.Sprintf("%s | *Build ID*: %s | %s | %s", r.getFrameworkName(), r.Metadata.Build, credentials.Get().Username, time.Now().Format("2006-01-02 15:04:05")), false, false)
+	contextSection := slack.NewSectionBlock(contextText, nil, nil)
+
+	resultText := slack.NewTextBlockObject("mrkdwn", r.GetRenderedResult(), false, false)
+	resultSection := slack.NewSectionBlock(resultText, nil, nil)
+
+	return []slack.Block{headerSection, contextSection, resultSection}
+}
+
+func (r *Reporter) getFrameworkName() string {
+	if r.Framework == "xcuitest" {
+		return "XCUITest"
+	}
+	if r.Framework == "testcafe" {
+		return "TestCafe"
+	}
+
+	return strings.Title(r.Framework)
+}
+
+func regenerateName(name, wholeName string, length int) string {
```

Comment:
A rather non-descriptive name for what the function does.
Looks like you are just adding spaces at the end? So essentially right-padding with spaces?
Would be good to name the function appropriately and also include a test for this function.

Reply:
I renamed this function and added a unit test.